buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java

```diff
@@ -217,22 +238,36 @@ public Configuration getExtracted() {
 
     @Override
     public TaskDependency getBuildDependencies() {
-        // For non-required Docker distributions, skip building the distribution is Docker is unavailable
-        if (isDocker() && getFailIfUnavailable() == false && dockerSupport.get().getDockerAvailability().isAvailable == false) {
-            return task -> Collections.emptySet();
-        }
+        Optional<TaskDependency> dockerBuildDependencies = dockerBuildDependencies();
+        return dockerBuildDependencies.orElseGet(() -> {
+            maybeFreeze();
+            return getType().shouldExtract() ? extracted.getBuildDependencies() : configuration.getBuildDependencies();
+        });
+    }
 
-        return configuration.getBuildDependencies();
+    public TaskDependency getArchiveBuildDependencies() {
+        Optional<TaskDependency> dockerBuildDependencies = dockerBuildDependencies();
+        return dockerBuildDependencies.orElseGet(() -> {
+            maybeFreeze();
+            return configuration.getBuildDependencies();
+        });
+    }
+
+    private Optional<TaskDependency> dockerBuildDependencies() {
+        // For non-required Docker distributions, skip building the distribution is Docker is unavailable
+        return (isDocker() && getFailIfUnavailable() == false && dockerSupport.get().getDockerAvailability().isAvailable == false)
```

Comment:
This doesn't seem right. In both scenarios we return what is effectively an empty set of dependencies. If docker is available, or we don't consider this to be optional, we should return the configuration's build dependencies.

Reply:
Are you saying that how we set this up currently on master is already wrong? https://github.com/elastic/elasticsearch/blob/7e5936d58ce0da5284fe95478b7ec4efe89e2180/buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java#L221-L223

Reply:
reworked this